### PR TITLE
[move 2024] Move mut var checks to cfgir

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- `mut` checks for locals did not consider if the local had not yet been assigned, e.g. `let x; x = 0;`
- The checks are now done during the locals safety checks 

## Test Plan 

- Updated tests

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [X] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

Improved Move's reporting of unused variables. In some cases, errors that were previously reported as unused assignments are now correctly marked as unused variables.

### CODE REVIEW

File: external-crates/move/crates/move-compiler-transactional-tests/tests/control_flow/named_block_nesting.move
Comment (cgswords) [inline external-crates/move/crates/move-compiler-transactional-tests/tests/control_flow/named_block_nesting.move:64]: I'm not sure whats going on here. Why did we lose this `mut`?

File: external-crates/move/crates/move-compiler/src/hlir/ast.rs
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/hlir/ast.rs:251]: I would have preferred having two different `Assign` commands, but I appreciate how much more work it would be to swap out everywhere.

File: external-crates/move/crates/move-compiler/tests/move_2024/migration/all_migrations.exp
Comment (cgswords) [inline external-crates/move/crates/move-compiler/tests/move_2024/migration/all_migrations.exp:29]: I believe this error needs to be nicer. This is an error from the following code, which should possibly live as its own test:

```
module a::m {
    struct S { f: u64 }

    fun t0(x: u64, s: S): u64 {
        let S { f: fin } = s;
        fin = 10;
        x = 20;
        fin + x
    }

}
```

I would expect mutability errors for both `fin` and `x`, but this error here is very strange. It's because the value for `fin` is dropped during reassignment, and it would be cool if we could get the error to explain that: renaming this as `_fin` is not really the correct fix.

File: external-crates/move/crates/move-compiler-transactional-tests/tests/control_flow/named_block_nesting.move
Comment (cgswords) [reply external-crates/move/crates/move-compiler-transactional-tests/tests/control_flow/named_block_nesting.move:64]: Oh, is it because these `i`s are never actually observed or used due to the break, so the assignment is discarded?

File: external-crates/move/crates/move-compiler/src/cfgir/liveness/mod.rs
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/cfgir/liveness/mod.rs:309]: A comment explaining the logic here may be nice: if this is an update, we know the variable isn't live after this, and so this assignment is subject to erasure.

File: external-crates/move/crates/move-compiler/src/cfgir/locals/mod.rs
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/cfgir/locals/mod.rs:294]: super nit: why move this?

File: external-crates/move/crates/move-compiler/src/expansion/ast.rs
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/expansion/ast.rs:268]: FWIW I think temps are subject to immutability. Match temps in the enum PR may be Either, though.

File: external-crates/move/crates/move-compiler/src/expansion/ast.rs
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/expansion/ast.rs:279]: Isn't mutability already an option? What is this option-option?

Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/expansion/ast.rs:279]: When would this be `None`?

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/parser/syntax.rs:2702]: nit: whitespace change

File: external-crates/move/crates/move-compiler/tests/move_2024/migration/let_mut_migrate.migration.exp
Comment (cgswords) [inline external-crates/move/crates/move-compiler/tests/move_2024/migration/let_mut_migrate.migration.exp:5]: While I appreciate the semantic correctness here, it is unnerving. :)

**Review by cgswords — COMMENTED** [review]
Left some comments. The only major one is about dropping values on accident and the weird error there.

File: external-crates/move/crates/move-compiler-transactional-tests/tests/control_flow/named_block_nesting.move
Comment (tnowacki) [reply external-crates/move/crates/move-compiler-transactional-tests/tests/control_flow/named_block_nesting.move:64]: Oh actually this one seems to be a bug, or something not really fixable because of the  unused assignment. I was just blindly changing it late at night. Let me take a deeper look

File: external-crates/move/crates/move-compiler/src/hlir/ast.rs
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/hlir/ast.rs:251]: I can make two different assign commands if you would prefer. But this seemed preferable not just because of the difficulty to refactor, but also since 90% of the time, you don't care about the case.

Again, happy to swap it out if you think it will be better long term

File: external-crates/move/crates/move-compiler/src/cfgir/locals/mod.rs
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/cfgir/locals/mod.rs:294]: Ah, yeah I needed it in a previous version of the PR. I'll revert

File: external-crates/move/crates/move-compiler/src/expansion/ast.rs
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/expansion/ast.rs:279]: This is None because we use Lvalue for assignment as well as bindings. There is no `mut` in the assignment case, and that is guaranteed by the grammar

File: external-crates/move/crates/move-compiler/tests/move_2024/migration/let_mut_migrate.migration.exp
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/tests/move_2024/migration/let_mut_migrate.migration.exp:5]: Wait what? This should hav emut still? Maybe an issue of unused bindings again?

File: external-crates/move/crates/move-compiler/tests/move_2024/migration/all_migrations.exp
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/tests/move_2024/migration/all_migrations.exp:29]: Fixed this, so its back to complaining about an unused assignment

**Comment by tnowacki** [comment]
I refactored the liveness code, which allows locals safety and borrow safety to handle unused variables a bit differently. I think it worked out quite nicely, and resulted in us reporting some unused variables that were previously missed

File: external-crates/move/crates/move-compiler/src/cfgir/locals/mod.rs
Comment (tzakian) [inline external-crates/move/crates/move-compiler/src/cfgir/locals/mod.rs:182]: Nit: spelling

**Review by tzakian — APPROVED** [review]
changes look good to me (and a super minor spelling nit).

**Comment by tnowacki** [comment]
Blocked by #16486

File: external-crates/move/crates/move-cli/tests/build_tests/migration/args.exp
Comment (cgswords) [inline external-crates/move/crates/move-cli/tests/build_tests/migration/args.exp:125]: Possibly left over from the other migration work?

File: external-crates/move/crates/move-cli/tests/build_tests/migration/Move.toml
Comment (cgswords) [inline external-crates/move/crates/move-cli/tests/build_tests/migration/Move.toml:13]: Possibly left over from migration work?

File: external-crates/move/crates/move-compiler/src/diagnostics/mod.rs
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/diagnostics/mod.rs:550]: 🎉

File: external-crates/move/crates/move-compiler/src/cfgir/liveness/mod.rs
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/cfgir/liveness/mod.rs:207]: Do we still need to take this argument?

**Review by cgswords — APPROVED** [review]
LGTM modulo a few small notes about the migration example (possibly a mistaken change).

I think in the future it might be nice to improve the unused assignment message to point to where the value is reassigned when it is unused due to reassignment before use, but we can always revisit this code to do that later.

**Pull request merged**
